### DISCUSSION TITLE
RestDocs와 Swagger의 장단점을 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#211] [Spring 뒷고기]
뭘까용?

**Comment by jimi567** [comment]
둘 모두 api 문서화를 자동화 하기 위해 사용하는 자동화 도구 입니다.

### Spring restDocs

+ 장점 : 테스트를 기반하여 문서화하기 때문에 테스트가 실패하게 된다면 문서화가 되지 않습니다. 따라서 정확성이 보장될 수 있습니다. Swagger와 비교해서 프로덕션 코드에 문서화에 필요한 코드가 녹아들지 않는다는 장점도 있습니다.

+ 단점 : 초기 설정이 굉장히 복잡합니다. 추가로 asciidoctor과 같은 추가 도구가 필요하게 됩니다. 
또 문서화를 위한 테스트도 추가적으로 작성해야 되고, 테스트를 통해 생성된 스니펫을 직접 개발자가 조합하여, html 문서를 생성하므로(자동화 방법이 있는지는 모르겠음)  문서화에 필요한 리소스가 많이 듭니다. 


### Swagger

+ 장점 :  코드 주석이나 어노테이션을 통해 api 문서를 쉽게 생성할 수 있습니다. Open Api Spec을 따르므로, 표준화된 형식으로 문서를 작성할 수 있습니다. 또 다른 시스템과의 통합 용이합니다.

+ 단점 : 프로덕션 코드에 문서화를 위한 코드가 녹아들게 됩니다. 이로 인해 프로덕션 코드의 가독성을 해칠 수 있습니다.  또 어노테이션 기반으로 문서가 자동 생성되므로  테스트가 적절하게 짜여져 있지 않다면, 실제 기대되는 동작과 문서가 일치하지 않을 수 있습니다.

**Comment by GIVEN53** [comment]
### Spring REST Docs
#### 장점
- 정확성: 실제 테스트 코드에 기반한 문서 생성으로 API 문서의 정확성을 보장합니다.
- 유연성: 다양한 템플릿 엔진(AsciiDoc, Markdown 등)을 지원하여 원하는 형태로 문서를 커스터마이징 할 수 있습니다.
- 개발 과정 통합: Spring 테스트 프레임워크와 통합되어 있어, 개발 과정 중 자연스럽게 문서가 생성됩니다.

#### 단점
- 러닝 커브: 설정과 사용법이 다소 복잡할 수 있으며, 특히 처음 사용하는 개발자에게는 학습이 필요합니다.
- 수동 작업 필요: 일부 문서화 작업은 수동으로 작성해야 합니다. (.adoc 결합)
- 비동기 API 문서화 어려움: 비동기 API에 대한 문서화를 설정하기가 상대적으로 어렵습니다.

### Swagger (OpenAPI)
#### 장점
- 간편함: 코드 주석을 기반으로 문서가 자동 생성되며, 추가적인 설정 없이도 바로 문서를 확인할 수 있습니다.
- 시각화 도구: Swagger UI를 통해 API를 시각적으로 표현하고, 바로 테스트할 수 있는 환경을 제공합니다.
- 생태계: Swagger/OpenAPI 생태계는 매우 넓고, 다양한 도구와 통합되어 있습니다.

#### 단점
- 정확성 문제: 코드 주석과 실제 구현이 일치하지 않을 경우, 문서의 정확성에 문제가 발생할 수 있습니다.
- 가독성: 프로덕션 코드에 코드 주석을 작성하기 때문에 가독성을 떨어뜨릴 수 있습니다.
- 유지보수: 구현이 변경될 경우 코드 주석의 변경 사항을 쉽게 파악하기 어려울 수 있습니다.

### 결론
Spring REST Docs는 테스트 기반 문서화를 원하는 팀에 적합하며, API 문서의 정확성과 일관성을 중시하는 경우에 유리합니다.
Swagger는 자동화된 문서화와 시각화 도구를 통해 빠르게 API 문서를 생성하고 공유하고자 하는 팀에 적합합니다.

프로젝트의 성격, 팀의 구성원 숙련도, 문서화 요구사항 등을 고려하여 두 도구 중 하나를 선택하는 것이 좋습니다. 경우에 따라 두 도구를 병행하여 사용하는 것도 가능합니다.
